### PR TITLE
done

### PR DESCRIPTION
**Pull request by norog6**

### CODE REVIEW

File: src/main/java/com/solbeg/Main.java [hunk 1]
```diff
@@ -76,7 +79,12 @@ public static void processUser(UserProvider userProvider, UserService userServic
      * @return provided or generated user
      */
     public static User getOrGenerateUser(UserProvider userProvider) {
-        throw new RuntimeException();
+        Optional<User> optionalUser = userProvider.getUser();
+        if (optionalUser.isPresent()) {
+            return optionalUser.get();
+        } else {
```
Comment (MusienkoYuriy97) [inline src/main/java/com/solbeg/Main.java:85]: remove if else
use orElseGet

File: src/main/java/com/solbeg/Main.java [hunk 2]
```diff
@@ -107,7 +116,7 @@ public static User getUser(UserProvider userProvider) {
      * @return optional credit balance
      */
     public static Optional<BigDecimal> retrieveCreditBalance(UserBankAccountProvider userBankAccountProvider) {
-        throw new RuntimeException();
+        return userBankAccountProvider.getUserBankAccount().map(UserBankAccount::getCreditBalance).get();
```
Comment (MusienkoYuriy97) [inline src/main/java/com/solbeg/Main.java:119]: use flatMap
userBankAccountProvider.getUserBankAccount().flatMap(UserBankAccount::getCreditBalance);
https://stackoverflow.com/questions/30864583/what-is-the-difference-between-optional-flatmap-and-optional-map

File: src/main/java/com/solbeg/Main.java [hunk 3]
```diff
@@ -129,7 +138,18 @@ public static Optional<User> retrieveUserGmail(UserProvider userProvider) {
...
+            return optionalUser.get();
+        } else {
+            Optional<User> fallbackUser = fallbackProvider.getUser();
+            if (fallbackUser.isPresent()) {
+                return fallbackUser.get();
+
+            } else {
+                throw new NoSuchElementException("No User provided by both providers!");
+            }
+        }
```
Comment (MusienkoYuriy97) [inline src/main/java/com/solbeg/Main.java:152]: here you can use or()
return userProvider.getUser().or(fallbackProvider::getUser).orElseThrow(NoSuchElementException::new);

File: src/main/java/com/solbeg/Main.java [hunk 4]
```diff
@@ -32,7 +30,8 @@ public static Optional<String> optionalOfString(String text) {
      * @param amount       money to deposit
      */
     public static void deposit(UserProvider userProvider, BigDecimal amount) {
-        throw new RuntimeException();
+
+        userProvider.getUser().ifPresent(x->x.setBalance(x.getBalance().add(amount)));
```
Comment (MusienkoYuriy97) [inline src/main/java/com/solbeg/Main.java:34]: Bad naming
replace "x" with "user"

File: src/main/java/com/solbeg/Main.java [hunk 5]
```diff
@@ -86,7 +94,8 @@ public static User getOrGenerateUser(UserProvider userProvider) {
      * @return optional balance
      */
     public static Optional<BigDecimal> retrieveBalance(UserProvider userProvider) {
-        throw new RuntimeException();
+        Optional<User> optionalUser = userProvider.getUser();
+        return optionalUser.map(User::getBalance);
```
Comment (MusienkoYuriy97) [inline src/main/java/com/solbeg/Main.java:98]: collapse it into 1 line 
return userProvider.getUser().map(User::getBalance);

File: src/main/java/com/solbeg/Main.java (hunk 1 above)
Comment (norog6) [reply src/main/java/com/solbeg/Main.java:85]: fixed

File: src/main/java/com/solbeg/Main.java [hunk 6]
```diff
@@ -65,7 +64,11 @@ public static User getUser(UserProvider userProvider, User defaultUser) {
      * @param userService
      */
     public static void processUser(UserProvider userProvider, UserService userService) {
-        throw new RuntimeException();
+        if (userProvider.getUser().isPresent()) {
+            userService.processUser(userProvider.getUser().get());
+        } else {
+            userService.processWithNoUser();
+        }
```
[the remark this answers is not in the document]
Comment (norog6) [reply src/main/java/com/solbeg/Main.java:71]: fixed

File: src/main/java/com/solbeg/Main.java (hunk 2 above)
Comment (norog6) [reply src/main/java/com/solbeg/Main.java:119]: fixed

File: src/main/java/com/solbeg/Main.java (hunk 3 above)
Comment (norog6) [reply src/main/java/com/solbeg/Main.java:152]: fixed

File: src/main/java/com/solbeg/Main.java (hunk 5 above)
Comment (norog6) [reply src/main/java/com/solbeg/Main.java:98]: fixed

File: src/main/java/com/solbeg/Main.java (hunk 4 above)
Comment (norog6) [reply src/main/java/com/solbeg/Main.java:34]: fixed

File: src/main/java/com/solbeg/Main.java [hunk 7]
```diff
@@ -161,6 +184,6 @@ public static OptionalDouble findMinBalanceValue(List<User> users) {
      * @return total credit balance
      */
     public static double calculateTotalCreditBalance(List<UserBankAccount> bankAccounts) {
-        throw new RuntimeException();
+        return bankAccounts.stream().mapToDouble(account->account.getCreditBalance().get().doubleValue()).sum();
```
[the remark this answers is not in the document]
Comment (norog6) [reply src/main/java/com/solbeg/Main.java:187]: fixed

File: src/main/java/com/solbeg/Main.java [hunk 8]
```diff
@@ -65,7 +64,8 @@ public static User getUser(UserProvider userProvider, User defaultUser) {
      * @param userService
      */
     public static void processUser(UserProvider userProvider, UserService userService) {
-        throw new RuntimeException();
+         userProvider.getUser().ifPresentOrElse(user->userService.processUser(user),
+                 ()->userService.processWithNoUser());
```
Comment (MusienkoYuriy97) [inline src/main/java/com/solbeg/Main.java:68]: replace with userService::processUser, userService::processWithNoUser

File: src/main/java/com/solbeg/Main.java [hunk 9]
```diff
@@ -150,7 +154,8 @@ public static User getUserWithMaxBalance(List<User> users) {
      * @return the lowest balance values
      */
     public static OptionalDouble findMinBalanceValue(List<User> users) {
-        throw new RuntimeException();
+
```
Comment (MusienkoYuriy97) [inline src/main/java/com/solbeg/Main.java:157]: remove empty line

**Pull request merged**